apps/api/src/app/partner-integrations/usecases/setup-vercel-integration/setup-vercel-integration.usecase.ts

```diff
@@ -0,0 +1,155 @@
+    try {
+      const tokenData = await this.getVercelToken(command.vercelIntegrationCode);
+
+      const projects = await this.getVercelProjects(tokenData.accessToken, tokenData.teamId);
+
+      const projectIds = projects.map((project) => project.id);
+
+      const envKeys = await this.getEnvKeys(command);
+
+      await this.setEnvironments({
+        projectIds,
+        identifier: envKeys.clientKey,
+        secretKey: envKeys.privateKey,
+        teamId: tokenData.teamId,
+        token: tokenData.accessToken,
+      });
+
+      return {
+        success: true,
+      };
+    } catch (error) {
+      throw new ApiException(
+        error?.response?.data?.error_description || error?.response?.data?.message || error.message
+      );
+    }
+  }
+
+  async getVercelToken(code: string): Promise<{
+    accessToken: string;
+    userId: string;
+    teamId: string | null;
+  }> {
+    const postData = new URLSearchParams({
+      code,
+      client_id: process.env.VERCEL_CLIENT_ID,
+      client_secret: process.env.VERCEL_CLIENT_SECRET,
+      redirect_uri: process.env.VERCEL_REDIRECT_URI,
+    });
+
+    const response = await lastValueFrom(
+      this.httpService.post(`${process.env.VERCEL_BASE_URL}/v2/oauth/access_token`, postData, {
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+      })
+    );
+
+    const data = response.data;
+
+    return {
+      accessToken: data.access_token,
+      userId: data.user_id,
+      teamId: data.team_id,
+    };
+  }
+
+  async getVercelProjects(token: string, teamId: string | null) {
+    const response = await lastValueFrom(
+      this.httpService.get(`${process.env.VERCEL_BASE_URL}/v4/projects${teamId ? `?teamId=${teamId}` : ''}`, {
+        headers: {
+          Authorization: `Bearer ${token}`,
+        },
+      })
+    );
+
+    return response.data.projects;
+  }
+
+  async getEnvKeys(command: SetupVercelIntegrationCommand): Promise<{
+    privateKey: string;
+    clientKey: string;
+  }> {
+    const apiKeysResponse = await this.getApiKeyUsecase.execute(
+      GetApiKeysCommand.create({
+        environmentId: command.environmentId,
+        organizationId: command.organizationId,
+        userId: command.userId,
+      })
+    );
+
+    const envResponse = await this.getEnvironmentUsecase.execute(
+      GetEnvironmentCommand.create({
+        organizationId: command.organizationId,
+        userId: command.userId,
+        environmentId: command.environmentId,
+      })
+    );
+
+    return {
+      privateKey: apiKeysResponse[0].key,
+      clientKey: envResponse.identifier,
+    };
+  }
+
+  async setEnvironments({ identifier, projectIds, secretKey, teamId, token }: ISetEnvironment): Promise<void> {
+    const projectApiUrl = `${process.env.VERCEL_BASE_URL}/v9/projects`;
+    const target = ['production', 'preview', 'development'];
```

Comment:
I have a doubt. These targets are our environments our it would be Vercel environments?

Reply:
These are for the Vercel environments.

Reply:
> I personally struggle to understand what you are proposing as this params are needed for the signup and github links that are used in the component rendering of the LoginForm. Would you mind to clarify a bit what you are proposing? 🤔

 Vercel itself dynamically adds `code` and `next` params, we need `code` to communicate with the Vercel API and `next` to finish the installation.